File: BasicAlgorithms/Sorting/L2-merge-sort/merge-sort.py
my_array1 = [70, 30, 10, 50, 100, 2] #test case 1
my_array2 = [5,2,3,1] #test case 2
my_array3 = [45, 10, 63, 40, 50, 80] #test case 3


def merge_sort(nums):
    def split(arr):
        mid = len(arr) // 2
        left_array = arr[:mid]
        right_array = arr[mid:]
        return left_array, right_array


    def merge(left_array, right_array):
        arr = []
        j = 0
        i = 0

        while (i < len(left_array) and j < len(right_array)):
            if left_array[i] < right_array[j]:
                arr.append(left_array[i])
                i +=1
            else:
                arr.append(right_array[j])
                j +=1
            
        while i < len(left_array):
            arr.append(left_array[i])
            i +=1
            
        while j < len(right_array):
            arr.append(right_array[j])
            j +=1
        
        return arr

    if len(nums) <= 1:
        return nums
    
    left, right = split(nums)
    split_left = merge_sort(left)
    split_right = merge_sort(right)
    return merge(split_left, split_right)


print(merge_sort(my_array1))
print(merge_sort(my_array2))
print(merge_sort(my_array3))

File: content/BasicAlgorithms/Sorting/L2-merge-sort/merge-sort.js
/** */
const myArray1 = [70, 30, 10, 50, 100, 2] //test case 1
const myArray2 = [1, 2, 3, 4, 5, 6] //test case 2
const myArray3 = [45, 10, 63, 40, 50, 80] //test case 3

/**
 * @TimeComplexity - O(nlog N)
 * @SpaceComplexity - O(n)
 * @param {*} arr 
 * @returns - sorted list
 */
const merge_sort = (arr)=>{
    if (arr.length <= 1) {
        return arr
    }
    const mid = Math.floor(arr.length / 2)
    const leftArray = arr.slice(0, mid)
    const rightArray = arr.slice(mid)

    const sortedLeft = merge_sort(leftArray)
    const sortedRight = merge_sort(rightArray)

    const result = []
    let j = 0 
    let i = 0

    while (i < sortedLeft.length && j < sortedRight.length) {
        if (sortedLeft[i] < sortedRight[j]) {
            result.push(sortedLeft[i])
            i += 1
        } else {
            result.push(sortedRight[j])
            j += 1
        }
    }

    return result.concat(sortedLeft.slice(i).concat(sortedRight.slice(j)))
}

console.log(merge_sort(myArray1))
console.log(merge_sort(myArray2))
console.log(merge_sort(myArray3))